driver/src/orderbook/shadow_orderbook.rs

```diff
@@ -10,12 +10,140 @@
 use super::StableXOrderBookReading;
 use crate::models::{AccountState, Order, TokenId};
 use anyhow::Result;
-use ethcontract::Address;
+use ethcontract::{Address, U256};
 use std::collections::{HashMap, HashSet};
+use std::sync::mpsc::{self, Receiver, SyncSender};
+use std::thread::{self, JoinHandle};
 
 /// A type definition representing a complete orderbook.
 type Orderbook = (AccountState, Vec<Order>);
 
+/// A shadowed orderbook reader where two orderbook reading implementations
+/// compare results.
+pub struct ShadowedOrderbookReader<'a> {
+    primary: &'a dyn StableXOrderBookReading,
+    shadow_thread: JoinHandle<()>,
+    shadow_channel: SyncSender<(u32, Orderbook)>,
+}
+
+impl<'a> ShadowedOrderbookReader<'a> {
+    /// Create a new instance of a shadowed orderbook reader that starts a
+    /// background thread
+    pub fn new(
+        primary: &'a dyn StableXOrderBookReading,
+        shadow: Box<dyn StableXOrderBookReading + Send + 'static>,
+    ) -> Self {
+        // NOTE: Create a bounded channel with a 0-sized buffer, this makes it
+        //   if the primary orderbook is read and the shadow is still reading,
+        //   the diff for that specific orderbook is skipped.
+        let (shadow_channel_tx, shadow_channel_rx) = mpsc::sync_channel(0);
+        let shadow_thread =
+            thread::spawn(move || background_shadow_reader(&*shadow, shadow_channel_rx));
```

Comment:
Not necessary in this PR, but if I wanted to test this class, I would try to implement a small TaskQueue abstraction (maybe some lightweight implementation already exists) that can be initialized with thread/channels for actually running the program, but that can also have a fake synchronous implementation we it directly calls the lambda on `try_send`.

This would allow testing the logic of invoking the secondary reader once the primary completes without any thread/wait logic.

Reply:
>  I would try to implement a small TaskQueue abstraction 

Thats a neat idea, I'll give that a shot in a separate PR then.